### PR TITLE
KEYCLOAK-18370 Introduce QueryParameters

### PR DESCRIPTION
**Pull request by mhajas**

### CODE REVIEW

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java [hunk 1]
```diff
@@ -88,22 +93,35 @@ public long delete(ModelCriteriaBuilder<M> criteria) {
...
-        for (Iterator<Entry<K, V>> iterator = store.entrySet().iterator(); iterator.hasNext();) {
-            Entry<K, V> next = iterator.next();
-            if (keyFilter.test(next.getKey()) && entityFilter.test(next.getValue())) {
-                res++;
-                iterator.remove();
-            }
+        Stream<Entry<K, V>> storeStream = store.entrySet().stream();
+        final AtomicLong res = new AtomicLong(0);
+
+        if (queryParameters != null && queryParameters.getOrdering() != null && !queryParameters.getOrdering().isEmpty()) {
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java:99]: Would it be possible to avoid all of these `null` checks?

File: server-spi/src/main/java/org/keycloak/storage/SearchableModelField.java [hunk 2]
```diff
@@ -40,31 +40,6 @@ public String getName() {
...
-        if (getClass() != obj.getClass()) {
-            return false;
-        }
-        final SearchableModelField<?> other = (SearchableModelField<?>) obj;
-        if ( ! Objects.equals(this.name, other.name)) {
-            return false;
-        }
-        return true;
-    }
-
```
Comment (hmlnarik) [inline server-spi/src/main/java/org/keycloak/storage/SearchableModelField.java:67]: Why is this removed?

File: model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java [hunk 3]
```diff
@@ -291,8 +290,17 @@ public UserSessionModel getUserSession(RealmModel realm, String id) {
...
+
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, false)
+                .compare(UserSessionModel.SearchableFields.CLIENT_ID, ModelCriteriaBuilder.Operator.EQ, client.getId());
+
+
+        return userSessionTx.read(mcb, userSessionStore.createQueryParametersBuilder()
+                    .pagination(firstResult, maxResults, UserSessionModel.SearchableFields.LAST_SESSION_REFRESH)
+                    .build())
+                .map(userEntityToAdapterFunc(realm))
+                .filter(Objects::nonNull);
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java:303]: Should these two lines be swapped?

File: model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    public static class OrderingTuple<M> {
+        private final SearchableModelField<M> modelField;
+        private final Order order;
+
+        public OrderingTuple(SearchableModelField<M> modelField, Order order) {
+            this.modelField = modelField;
+            this.order = order;
+        }
+
+        public SearchableModelField<M> getModelField() {
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java:46]:
```suggestion
        public SearchableModelField<M> getField() {
```

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java [hunk 5]
```diff
@@ -68,18 +68,48 @@
...
+     *
+     * @param criteria Criteria filtering out the object, originally obtained
+     *   from {@link #createCriteriaBuilder()} method of this object.
+     *   If {@code null}, it returns an empty stream.
+     * @param queryParameters parameters for the query like firstResult, maxResult, requested ordering, etc.
+     * @return Stream of objects. Never returns {@code null}.
+     * @throws IllegalStateException If {@code criteria} is not compatible, i.e. has not been originally created
+     *   by the {@link #createCriteriaBuilder()} method of this object.
+     */
+    Stream<V> read(ModelCriteriaBuilder<M> criteria, QueryParameters<M> queryParameters);
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:87]: Would something like this be possible?
```java
import static QueryParameters.*;

read(withCriteria(mcb));
read(withCriteria(mcb).paginate(offset, limit, CLIENT_ID));
read(withCriteria(mcb).orderBy(ID));
```

**Review by hmlnarik — COMMENTED** [review]
Thanks @mhajas, it's a great start.

When going through the code, I found few places that could be perhaps improved in terms of usability and performance. It extends to the tree store as well. I'm attaching few of the ideas here.

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java [hunk 6]
```diff
@@ -121,6 +163,18 @@
...
+     * Returns query parameters builder for current storage.
+     * 
+     * Using {@code QueryParametersBuilder} it is possible to specify various additional parameters for queries
+     * like offset (how many objects to skip before returning the first one), limit (maximum number of objects to 
+     * return) or fields (in form of the corresponding {@link org.keycloak.storage.SearchableModelField}) that will be 
+     * used for ordering resulting streams of objects
+     * 
+     * @return See description. Never return {@code null}
+     */
+    QueryParameters.Builder<M> createQueryParametersBuilder();
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:176]: I expected a discussion around this method. I started removing it just before I sent the PR, but then I reverted. What it gives us, to have it this way, is casting to the correct type without any additional work. The `M` template is a model type in `MapStorage`. If we would want to have this directly created in the `*Providers` we would need to provide a model type each time we create it. So, for example, for users, we would need to call this 
```java
QueryParameters.Builder.<UserModel>create()
```

The correct type is necessary so that we have SearchableModelField checked for the correct model.

File: model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            return modelField;
+        }
+
+        public Order getOrder() {
+            return order;
+        }
+    }
+
+    public static class Builder<M> {
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java:55]: Do we want to have `QueryParameters` immutable? I thought it could save us some problems in the future, that is why I created the builder.

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:176]: One can get the right type with static methods as well, which could do the trick, see e.g. [Collections.emptySet()](https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#emptySet())

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java (hunk 1 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java:99]: It seems that this question is related to the first one regarding the default instance of `QueryParameters`. The question is whether we want to have some default comparator for each entity or we want to allow pagination without setting the `orderBy` attribute and rely only on the storage that each call will result in the same entity ordering and hence the pagination will be correct.

File: server-spi/src/main/java/org/keycloak/storage/SearchableModelField.java (hunk 2 above)
Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/storage/SearchableModelField.java:67]: Because of [this](https://github.com/mhajas/keycloak/blob/8feef482dc8400ba37aa128d101923d7b435eda0/model/map/src/main/java/org/keycloak/models/map/storage/MapFieldPredicates.java#L93): 

```java
private static final Map<SearchableModelField<?>, Comparator<?>> COMPARATORS = new HashMap<>();
```
and 
we have more `SearchableModelFields` from more models with the same name (for example CLIENT_ID).

File: model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java:55]: Not necessarily, we can make it constant

File: server-spi/src/main/java/org/keycloak/storage/SearchableModelField.java (hunk 2 above)
Comment (hmlnarik) [reply server-spi/src/main/java/org/keycloak/storage/SearchableModelField.java:67]: It seems we should come up with a better `equals` then :)

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java (hunk 5 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:87]: It would be possible but this would require probably some more discussion. 

This way we would have some object that is wrapping `QueryParameters` together with `ModelCriteriaBuilder`. Isn't this basically the same as having QueryParameters stored directly in the `ModelCriteriaBuilder`? 

It would end up like this:
```java
store.createCriteriaBuilder()
    .compare(aField, EQ, expectedValue)
    .pagination(first, max, orderByField)
```

File: server-spi/src/main/java/org/keycloak/storage/SearchableModelField.java (hunk 2 above)
Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/storage/SearchableModelField.java:67]: Do we need a custom `equals`? Do we plan to, for example, send a `SearchableField` instance to another Keycloak node? If not we can have the default `equals` because all fields are static objects so within one VM a searchableField will be always equal only to itself.

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:176]: The problem is caused by the usage of the builder. In some cases we do not put `QueryParameters.Builder` anywhere we just directly create `QueryParameters` from it, and it seems in such cases Java can't infer the type.

File: model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java:55]: What do you mean by constant? Do you mean that we need to initialize all parameters in the constructor. If yes then setting `orderBy` may be complicated. That is why I introduced Builder to make it simple like `.orderBy(1field).orderBy(2field, DESC).orderBy(3field, DESC)` etc. 

We can, of course, do without Builder I just want to understand why we don't want it.

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java (hunk 5 above)
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:87]: The pagination must not be available for individual criteria. What would be the meaning of
```java
  store.createCriteriaBuilder()
    .or(store.createCriteriaBuilder() .compare(aField, EQ, expectedValue) .pagination(first, max, orderByField),
         store.createCriteriaBuilder() .compare(aField2, EQ, expectedValue2) .pagination(first, max, orderByField2))
    .pagination(0, 1)
```
?

Pagination and order is _query_ parameter, not criteria parameter. Thus it cannot be incorporated in the `ModelCriteriaBuilder`.

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:176]: Could you provide an example?

File: model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java:55]: My point was slightly elsewhere:

The `Builder` and all of the calls to `store.getQueryParametersBuilder().[...].build()` are a boilerplate code that should not be necessary. Ideally, the whole part would be short, like `withCriteria` or `orderBy` above which makes the code more readable and maintainable. Also, there is nothing storage-specific in the query parameters (it's all about model fields and numbers), so there is no need to have `MapStorage` implement another method. Sidenote - if we find the way to do it reasonably with query parameters, this can lead to refactor of `ModelCriteriaBuilder` use within `MapStorage` as well.

> What do you mean by constant?

Apologies, I should have explained precisely. By a constant I mean `public static final` field. Immutability is not strictly necessary (even though it would be nice).

> I just want to understand why we don't want it.

What is the maximum level of `orderBy` that we really need to support Keycloak? With order, we don't need to be fully generic, it is enough if we support the reasonable usecases.
Builder is excellent pattern for initialization of an immutable object with many fields. Neither pagination nor ordering has that many fields so usage of builder pattern rather obscures the use.

File: server-spi/src/main/java/org/keycloak/storage/SearchableModelField.java (hunk 2 above)
Comment (hmlnarik) [reply server-spi/src/main/java/org/keycloak/storage/SearchableModelField.java:67]: Thanks, you are right. In case of e.g. Infinispan, the infinispan representation of predicate would be generated from this representation within the same JVM, and only that representation (not the original model fields) would be sent over the wire.

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java (hunk 1 above)
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java:99]: Exactly:
- There should be a default pagination with a given maximum number of objects returned by a query, say 100 objects.
- The default order would be empty.
The providers that leverage this pagination should set the order explicitly if there should be one.

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:176]: Sure, look at following commit: https://github.com/mhajas/keycloak/commit/66007936937e49d2283b71f20c03fb08743d4b70

The build is failing with:
```
 /home/runner/work/keycloak/keycloak/model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:[588,74] incompatible types: org.keycloak.storage.SearchableModelField<org.keycloak.models.UserModel> cannot be converted to org.keycloak.storage.SearchableModelField<java.lang.Object>`
```

File: model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java (hunk 3 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java:303]: Probably no. This is there because the `userEntityToAdapterFunc` lambda can return `null` when user session is expired. This is basically copy&paste from other methods that are not using pagination.

File: model/map/src/main/java/org/keycloak/models/map/storage/MapKeycloakTransaction.java [hunk 8]
```diff
@@ -61,8 +61,20 @@
      * @param mcb criteria to filter values
      * @return values that fulfill the given criteria, that are updated based on changes in the current transaction
      */
-    Stream<V> read(ModelCriteriaBuilder<M> mcb);
+    default Stream<V> read(ModelCriteriaBuilder<M> mcb) {
+        return read(QueryParameters.withCriteria(mcb));
+    }
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/MapKeycloakTransaction.java:66]: Remove this one

File: model/map/src/main/java/org/keycloak/models/map/storage/MapKeycloakTransaction.java [hunk 9]
```diff
@@ -71,7 +83,18 @@
      * @param mcb criteria to filter values
      * @return number of values present in the storage that fulfill the given criteria
      */
-    long getCount(ModelCriteriaBuilder<M> mcb);
+    default long getCount(ModelCriteriaBuilder<M> mcb) {
+        return getCount(QueryParameters.withCriteria(mcb));
+    }
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/MapKeycloakTransaction.java:88]: Remove this

File: model/map/src/main/java/org/keycloak/models/map/storage/MapKeycloakTransaction.java [hunk 10]
```diff
@@ -117,7 +140,20 @@ default V getUpdated(V orig) {
      * @param artificialKey key to record the transaction with, must be a key that does not exist in this transaction to
      *                      prevent collisions with other operations in this transaction
      * @param mcb criteria to delete values
+     * @return number of removed objects (might return {@code -1} if not supported)
+     */
+    default long delete(K artificialKey, ModelCriteriaBuilder<M> mcb) {
+        return delete(artificialKey, QueryParameters.withCriteria(mcb));
+    }
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/MapKeycloakTransaction.java:147]: Remove this

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java [hunk 11]
```diff
@@ -69,18 +71,45 @@
      * @throws IllegalStateException If {@code criteria} is not compatible, i.e. has not been originally created
      *   by the {@link #createCriteriaBuilder()} method of this object.
      */
-    Stream<V> read(ModelCriteriaBuilder<M> criteria);
+    default Stream<V> read(ModelCriteriaBuilder<M> criteria) {
+        return read(withCriteria(criteria));
+    }
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:76]: Remove this

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java [hunk 12]
```diff
@@ -105,7 +134,18 @@
      * @throws IllegalStateException If {@code criteria} is not compatible, i.e. has not been originally created
      *   by the {@link #createCriteriaBuilder()} method of this object.
      */
-    long delete(ModelCriteriaBuilder<M> criteria);
+    default long delete(ModelCriteriaBuilder<M> criteria) {
+        return delete(withCriteria(criteria));
+    }
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:139]: Remove this

File: model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java [hunk 13]
```diff
@@ -69,18 +71,45 @@
...
+     * The criteria are specified in the given criteria builder based on model properties.
+     *
+     * @param criteria
+     * @return Number of objects. Never returns {@code null}.
+     * @throws IllegalStateException If {@code criteria} is not compatible, i.e. has not been originally created
+     *   by the {@link #createCriteriaBuilder()} method of this object.
+     */
+    default long getCount(ModelCriteriaBuilder<M> criteria) {
+        return getCount(criteria, null);
+    }
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java:100]: Remove this

**Review by hmlnarik — COMMENTED** [review]
Looks good, thanks! Since map module is an experimental module so far, I suggest removing the `ModelCriteriaBuilder`-based methods in `MapStorage` interface replaced by `QueryParameters` straightaway.

File: model/map/src/main/java/org/keycloak/models/map/authorization/MapPolicyStore.java [hunk 14]
```diff
@@ -111,8 +111,8 @@ public void delete(String id) {
     public Policy findById(String id, String resourceServerId) {
         LOG.tracef("findById(%s, %s)%s", id, resourceServerId, getShortStackTrace());
 
-        return tx.read(forResourceServer(resourceServerId)
-                .compare(SearchableFields.ID, Operator.EQ, id))
+        return tx.read(withCriteria(forResourceServer(resourceServerId)
+                .compare(SearchableFields.ID, Operator.EQ, id)))
                 .findFirst()
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/authorization/MapPolicyStore.java:116]: @hmlnarik I am just looking into this piece of code and it seems we should add the pagination to places like this. I think, we should go through the source code and optimize with pagination where possible. WDYT? And also should we do it as part of this PR or create and enhancement Jira in the cleanup epic? I would probably prefer the second option as this PR is already quite big.

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java [hunk 15]
```diff
@@ -88,15 +94,21 @@ public long delete(ModelCriteriaBuilder<M> criteria) {
...
+        final AtomicLong res = new AtomicLong(0);
+
+        if (!queryParameters.getOrderBy().isEmpty()) {
+            Comparator<V> comparator = MapFieldPredicates.getComparator(queryParameters.getOrderBy().stream());
+            storeStream = storeStream.sorted((entry1, entry2) -> comparator.compare(entry1.getValue(), entry2.getValue()));
         }
-        return res;
+
+        paginatedStream(storeStream.filter(next -> keyFilter.test(next.getKey()) && entityFilter.test(next.getValue()))
+                , queryParameters.getOffset(), queryParameters.getLimit())
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java:106]: @hmlnarik Does it make sense to have `limit` set also for delete method? This can be causing some problems in the future when we, for example, remove a realm with many clients and only 100 (default limit set in QueryParameters) is removed. Maybe we should ignore pagination for deletes completely as currently we don't manually set it anywhere in the code, and default limit 100 may cause problems. WDYT?

Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java:106]: @hmlnarik Bad news, I am not sure currently, we can set the default limit for each query to 100. There were some test failures with it. The problem is, that we do something like this:
```
policyStore.findByResourceServer(id).stream()
            .map(Policy::getId)
            .forEach(policyStore::delete);
```

This code is executed when you remove a resource server (for example when disabling authz for a client). If the client has more than 100 policies, there are some policies left in the database. This can be fixed however, I would again do it as a follow up enhancement including setting some default value for the `limit` parameter. WDYT?

File: model/map/src/main/java/org/keycloak/models/map/authorization/MapPolicyStore.java (hunk 14 above)
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/authorization/MapPolicyStore.java:116]: +1 for follow-up JIRA. It seems to be nice to have rather than must have, right?

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java (hunk 15 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java:106]: The problem is, that in this case, we are reading from the storage and then remove each policy we found. We would need to find all places like this and replace them with bulk deletes. But I am not sure it will be possible in each case as it would maybe require some join queries to achieve this.

Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java:106]: Looking at this for the second time: I would argue that this is a misuse of reading - there should be a method "delete all policies belonging to given resource server" in the storage, e.g. `policyStore.deleteByResourceServer(id)`. This way, even caching would be able to invalidate the right events.
I get your point though. How many places like this are in the codebase? It seems that they are located to fine-grained permissions?

Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java:106]: This is a good question. I don't know the answer yet. Potentially a lot, maybe only a few. I quickly looked at some occurences and clients/roles suffer from the same thing, so this is not only authz:
https://github.com/keycloak/keycloak/blob/a8fdd79d1b9d78e94a7025e2a3f367561fd7557d/model/map/src/main/java/org/keycloak/models/map/client/MapClientProvider.java#L177-L184

I would suggest to merge this as is, without the default value 100 for limit (current state of this PR), and add all discussed things in this PR in a separate Jira. The advantage would be, that all work that follows on this (Jpa, HotRod) can use changes from this PR immediately and these enhancements can be added anytime later as there won't be many conflicts probably. WDYT?

The other option is to address it in this PR, however I am not sure how much time I will be able to spend on this as I need to investigate the HotRod stuff.

File: model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+     * Sets pagination (offset, limit and orderBy) parameters to {@code QueryParameters}
+     *
+     * @param offset
+     * @param limit
+     * @param orderBy
+     * @return this object
+     */
+    public QueryParameters<M> pagination(Integer offset, Integer limit, OrderBy<M> orderBy) {
+        return pagination(offset, limit, Collections.singletonList(orderBy));
+    }
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java:66]: Are these methods used at all?

File: model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+
+    /**
+     * Sets orderBy parameter; can be called repeatedly; fields are stored in a list where the first field has highest
+     * priority when determining order; e.g. the second field is compared only when values for the first field are equal
+     *
+     * @param searchableModelField
+     * @return this object
+     */
+    public QueryParameters<M> orderBy(SearchableModelField<M> searchableModelField) {
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java:87]:
```suggestion
    public QueryParameters<M> orderBy(SearchableModelField<M> orderByAscField) {
```

**Comment by mhajas** [comment]
@hmlnarik Thank you very much for your review. I updated the PR with your latest suggestions. I also created [a new Jira](https://issues.redhat.com/browse/KEYCLOAK-18803) for the follow-up task we discussed. Feel free to add/update details.

**Review by hmlnarik — APPROVED** [review]
